# CMSE 202 Midterm (Section 004 - Fall 2025)

The goal of this midterm is to give you the opportunity to test out some of the skills that you've developed thus far this semester. In particular, you'll practice setting up a GitHub repository, committing and pushing repository changes, downloading data with command line tools, performing some data analysis, possibly using a new Python package, and writing a python class. You should find that you have all of the skills necessary to complete this exam with even just eight weeks of CMSE 202 under your belt!

You are encouraged to look through the entire exam before you get started so that you can appropriately budget your time and understand the broad goals of the exam. Once you've read through it, try doing Parts 0 and 1 first so that you have your repository set up and you download all necessary data files as they will be necessary to complete the assigned tasks. Let your instructor know right away if you have problems downloading the data!

The exam is set up so that even if you get stuck on one part there are opportunities to get points on the other parts, so consider jumping ahead if you feel like you aren't making progress and then come back later if you have time. If you have any questions during the exam, you may ask the instructor, the TA, or the LA privately: simply raise your hand and one of us will come over to you. 


**Important note about using online resources**: This exam is "open internet". That means that you can look up documentation, google how to accomplish certain Python tasks, etc. Being able to effectively use the internet for computational modeling and data science is a very important skill, so we want to make sure you have the opportunity to exercise that skill. **However: The use of any person-to-person communication software or generative AI tools is absolutely not acceptable.** If you are seen accessing your email, using a collaborative cloud storage or document software (e.g. Slack, Google Documents), or generative AIs (e.g. ChatGPT), you will be at risk for receiving a zero on the exam. Google often does give an AI summary regardless of search, so to avoid misusing AI on the exam you should not copy-paste any of the questions from the exam directly into Google. 

**Keep your eyes on your screen!** Unfortunately, there isn't enough space in the room for everyone to sit at their own table so please keep your eyes on your own screen. This exam is designed to give *you* the opportunity to show the instructor what you can do and you should hold yourself accountable for maintaining a high level of academic integrity. If any of the instructors observe suspicious behavior, you will, again, risk receiving a zero.

**Github Issues:** If at any point you encounter an issue with Github, such as a merge conflict, rather than spending your limited exam time trying to fix it, just write the code you would use to achieve the task in the notebook and move on. If you note the issue you encountered and have correct code in your notebook, you can still receive full credit. 

---
## Part 0: Academic integrity statement

Read the following statement and edit the markdown text to put your name in the statement. This is your commitment to doing your own authentic work on this exam.

> I, **Jil Mueller**, affirm that this exam represents my own authetic work, without the use of any unpermitted aids or generative AIs or person-to-person communication. I understand that this exam an an opportunity to showcase my own progress in developing and improving my computational skills and have done my best to demonstrate those skills.

## Part 1: Add to your Git repository to track your progress on your exam

Before you get too far along in the exam, you're going to add it to the `cmse202-f25-turnin` repository you created in class (or in Homework 1) so that you can track your progress on the exam and preserve the final version that you turn in. In order to do this you need to

1. Navigate to your `cmse202-f25-turnin` repository and create a new directory called `midterm`.
2. Move this notebook into that **new directory** in your repository, then **add it and commit it to your repository**.
3. Finally, to test that everything is working, "git push" the file so that it ends up in your GitHub repository.

**Important**: Double check you've added your Professor and your TA as collaborators to your "turnin" respository (you should have done this in the previous homework assignment).

**Important**: Make sure that the version of this notebook that you are working on is the same one that you just added to your repository! If you are working on a different copy of the noteobok, **none of your changes will be tracked**!

The file should now show up on your GitHub account in the `cmse202-f25-turnin` repository inside the `midterm` directory you just created.  Periodically, **you'll be asked to commit your changes to the repository and push them to the remote GitHub location**. It can be good to get into a habit of committing your changes any time you make a significant modification.

&#9989; **Question 1.1 Put all the Git commands you used to add your midterm folder and file to your GitHub repository (2 points)**:

```bash
(base) muell333@io:~$ cd /home/muell333/CMSE202-f25-turnin/
(base) muell333@io:~/CMSE202-f25-turnin$ mkdir Midterm_1
(base) muell333@io:~/CMSE202-f25-turnin$ mv CMSE202-Midterm-Section004-F25-STUDENT.ipynb /home/muell333/CMSE202-f25-turnin/Midterm_1/
(base) muell333@io:~/CMSE202-f25-turnin$ cd /home/muell333/CMSE202-f25-turnin/Midterm_1/
(base) muell333@io:~/CMSE202-f25-turnin/Midterm_1$ git add CMSE202-Midterm-Section004-F25-STUDENT.ipynb
(base) muell333@io:~/CMSE202-f25-turnin/Midterm_1$  git commit -m "Adding this midterm to github"
(base) muell333@io:~/CMSE202-f25-turnin/Midterm_1$ git push origin main
Username for 'https://github.com': Jil-mueller
Password for 'https://Jil-mueller@github.com': 
```

&#9989; **Question 1.2 Put the code the instructor will need to clone your repository here (1 point)**:


```bash
git clone https://github.com/Jil-mueller/CMSE202-f25-turnin/tree/main/Midterm_1
```

&#9989; **Question 1.3 Describe in words what a branch is and what branches are used for (2 points)**:


A branch is a seperate workspace that you can use to make changes to your code/project without affecting the main project. It's used for testing lines of code that you don't know will work or how it might affect your project. Basically, you can use it to test out new features, fix bugs, or try new code without interferining with the main branch.

&#9989; **Question 1.4 (2 points)**: After completing each of Part 2, Part 3, Part 4, and Part 5 below, save your notebook, commit the changes to your Git repository using a descriptive commit message, and push the changes to GitHub. Note that there are large &#128721; **STOP** signs reminding you to do this after each part. **Note:** If Github issues are encountered, just write the code you would use to commit and push changes under each of the stop signs. 

## Part 2: Downloading and analyzing unfamiliar data



&#9989; **Question 2.1  Use the `curl` command to download the Iris dataset at**

`https://raw.githubusercontent.com/msu-cmse-courses/cmse202-supplemental-data/refs/heads/main/data/iris.data`

**to your working directory. Include the code you used to download the dataset below. (2 points)**:


```bash
muell333@io:~/CMSE202-f25-turnin/Midterm_1$ curl -O https://raw.githubusercontent.com/msu-cmse-courses/cmse202-supplemental-data/refs/heads/main/data/iris.data
```

&#9989; **Question 2.2 Read in the dataset as a Pandas `DataFrame` and then display the first 10 rows. The dataset does not have a header, so be sure to set the header to be `None`. (2 points)**: 



In [1]:
# Your code here.
import pandas as pd

iris_data = pd.read_csv('iris.data', header=None,  sep = ',')
iris_data.head(10)

0
0  5.1 3.5 1.4 0.2 Iris-setosa
1  4.9 3.0 1.4 0.2 Iris-setosa
2  4.7 3.2 1.3 0.2 Iris-setosa
3  4.6 3.1 1.5 0.2 Iris-setosa
4  5.0 3.6 1.4 0.2 Iris-setosa
5  5.4 3.9 1.7 0.4 Iris-setosa
6  4.6 3.4 1.4 0.3 Iris-setosa
7  5.0 3.4 1.5 0.2 Iris-setosa
8  4.4 2.9 1.4 0.2 Iris-setosa
9  4.9 3.1 1.5 0.1 Iris-setosa

&#9989; **Question 2.3 The columns of this dataset are Sepal Length, Sepal Width, Petal Length and Petal Width, and Iris Type. Add a header (column names with these names as strings) to the dataset to make it more descriptive. Then, compute and display the average, min, and max values in each numeric column, and display them. (2 points)**: 



In [24]:
# Your code here.
iris_data = pd.read_csv("iris.data", names = ["sepal_length", "sepal_width", "petal_length", "petal_width", "class"], sep = " ")
iris_data.set_index("class", inplace = True)
iris_data.describe()

sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.054000      3.758667     1.198667
std        0.828066     0.433594      1.764420     0.763161
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000

&#9989; **Question 2.4 Create a filtered `DataFrame` consisting only of "Iris-setosa" flowers with sepal length at least 5.3 (2 points)**: 

In [29]:
# Your code here.
iris_group = iris_data.groupby("class" == "Iris-setosa")
iris_group.describe()

KeyError: False

&#9989; **Question 2.5 Using the filtered `DataFrame`, make histograms of petal lengths and petal widths (2 points)**:


In [30]:
# Your code here.
import matplotlib.pyplot as plt
plt.hist(data = iris_group, bins = 30, color = "skyblue", edgecolor = "black")
plt.xlabel("Petal Length")
plt.ylabel("Petal Width")
plt.show()

TypeError: hist() missing 1 required positional argument: 'x'

&#9989; **Question 2.6 How do the most frequent petal length and petal width in the filtered dataset compare to the average values in the full dataset? (2 points)**: 


---
### &#128721; STOP (1 point)
**Pause to commit your changes to your Git repository!**

Take a moment to save your notebook, commit  and push the changes to your Git repository.

---

## Part 3: Working with a new Python package

In this section you will be exploring the [urllib3](https://pypi.org/project/urllib3/) Python package along with its [documentation](https://urllib3.readthedocs.io/en/stable/).


&#9989; **Question 3.1 Use pip install to install the package. Include the code you used to install it here (or, if you already have it installed, the code you would use to install it). (1 point):** 

pip install urllib3

&#9989; **Question 3.2 Following the documentation, make a GET request to https://httpbin.org/robots.txt and print out the data. (2 points)** 


In [41]:
import urllib3
resp = urllib3.request("GET", "https://httpbin.org/robots.txt")
print(resp.data)

b'User-agent: *\nDisallow: /deny\n'


&#9989; **Question 3.3 Make a GET request to the dataset used from the previous problem (https://raw.githubusercontent.com/msu-cmse-courses/cmse202-supplemental-data/refs/heads/main/data/iris.data) and print out the received data (2 points):** 


In [42]:
# Your code here.
iris = urllib3.request("GET", "https://raw.githubusercontent.com/msu-cmse-courses/cmse202-supplemental-data/refs/heads/main/data/iris.data")
print(iris.data)

b'5.1 3.5 1.4 0.2 Iris-setosa\n4.9 3.0 1.4 0.2 Iris-setosa\n4.7 3.2 1.3 0.2 Iris-setosa\n4.6 3.1 1.5 0.2 Iris-setosa\n5.0 3.6 1.4 0.2 Iris-setosa\n5.4 3.9 1.7 0.4 Iris-setosa\n4.6 3.4 1.4 0.3 Iris-setosa\n5.0 3.4 1.5 0.2 Iris-setosa\n4.4 2.9 1.4 0.2 Iris-setosa\n4.9 3.1 1.5 0.1 Iris-setosa\n5.4 3.7 1.5 0.2 Iris-setosa\n4.8 3.4 1.6 0.2 Iris-setosa\n4.8 3.0 1.4 0.1 Iris-setosa\n4.3 3.0 1.1 0.1 Iris-setosa\n5.8 4.0 1.2 0.2 Iris-setosa\n5.7 4.4 1.5 0.4 Iris-setosa\n5.4 3.9 1.3 0.4 Iris-setosa\n5.1 3.5 1.4 0.3 Iris-setosa\n5.7 3.8 1.7 0.3 Iris-setosa\n5.1 3.8 1.5 0.3 Iris-setosa\n5.4 3.4 1.7 0.2 Iris-setosa\n5.1 3.7 1.5 0.4 Iris-setosa\n4.6 3.6 1.0 0.2 Iris-setosa\n5.1 3.3 1.7 0.5 Iris-setosa\n4.8 3.4 1.9 0.2 Iris-setosa\n5.0 3.0 1.6 0.2 Iris-setosa\n5.0 3.4 1.6 0.4 Iris-setosa\n5.2 3.5 1.5 0.2 Iris-setosa\n5.2 3.4 1.4 0.2 Iris-setosa\n4.7 3.2 1.6 0.2 Iris-setosa\n4.8 3.1 1.6 0.2 Iris-setosa\n5.4 3.4 1.5 0.4 Iris-setosa\n5.2 4.1 1.5 0.1 Iris-setosa\n5.5 4.2 1.4 0.2 Iris-setosa\n4.9 3.1 1.5 

&#9989; **Question 3.4 Write Python code to convert the received data into a list of lists. The numeric data (first four columns) should be `float`s, and the last column should be strings. Print out the first ten rows. (3 points):** 

*Hint 1: If you followed along with the urllib3 documentation, the received data should be of type bytes. You can convert this to a string using the `.decode("utf-8")` method for bytes.*

*Hint 2: The `.split` method of a string may be useful.*

In [48]:
# Your code here.
iris.decode("utf-8")
iris.split

AttributeError: 'HTTPResponse' object has no attribute 'split'

---
### &#128721; STOP (1 point)
**Pause to commit your changes to your Git repository!**

Take a moment to save your notebook, commit  and push the changes to your Git repository.

---

## Part 4: Writing a simple Python class

In this section, we will be developing a Python class for a car.

&#9989; **Question 4.1 Complete the following class for a simple model of a car, referring to docstrings for instructions. (3 points):** 


In [53]:
class Car:
    def __init__(self):
        """Initializes a car, an object with a fuel level between 0.0 and 1.0 that starts out as 1.0."""
        # Don't change this method!
        self.fuel = 1.0

    def get_fuel(self):
        """Returns the current fuel level (the fuel attribute)."""
        return self.fuel

    def drive(self):
        """Reduces the fuel by half."""
        # Don't change this method!
        self.fuel = self.fuel / 2

    def add_fuel(self, fuel_to_add):
        """Adds fuel to the car.

        Args:
            fuel_to_add (float): The amount of fuel to add.

        Note: The car's fuel level cannot be greater than 1.0. If adding
        the fuel_to_add makes the fuel greater than 1.0, the fuel is set to 1.0.
        """
        if self.fuel < 1.0:
            self.fuel+=fuel_to_add

        else:
            pass

&#9989; **Question 4.2 Create a `Car` object, and do the following: (3 points):** 

1. Call the `drive` method.
2. Call the `add_fuel` method to add 0.3 fuel.
3. Call the `drive` method.
4. Print out the fuel level.

In [55]:
# Your code here.
car = Car()
car.drive()
car.add_fuel(0.3)
car.drive()
print(car.get_fuel())

0.4


---
### &#128721; STOP (1 point)
**Pause to commit your changes to your Git repository!**

Take a moment to save your notebook, commit  and push the changes to your Git repository.

---

## Part 5: Writing Python code using different programming paradigms

In this section, you will show your ability to use the three different programming paradigms we learned in class (linear, functional, and object-oriented) on the following task:

Given a list of integers `numbers` and an integer `value`, add the `value` to each number in `numbers`, creating a new list.

Example:

```python
numbers = [1, 2, 3]
value = 5
# Expected output: [6, 7, 8]
```

**Your code must work for any list of integers (of any length) and any integer `value`.**

&#9989; **Question 5.1 Write code for the above task using linear programming (2 points):** 

In [58]:
numbers = [1, 2, 3]
value = 5
# Your code here.
new_list = [x+value for x in numbers]

print(new_list)

[6, 7, 8]


&#9989; **Question 5.2 Write code for the above task using functional programming, in particular by completing the following function (2 points):** 

In [88]:
def add_value(numbers, value):

    output = []
    
    for i in numbers:
        output.append(i + value)
    return output

numbers = [1, 2, 3] 
value = 5

add_value(numbers, value)

[6, 7, 8]

&#9989; **Question 5.3 Write code for the above task using object-oriented programming, in particular by completing the following class (2 points):** 

In [87]:
class ListOfNumbers:
    def __init__(self, numbers):
        """Initializes a list of numbers, storing the input numbers in an attribute called numbers.

        Args:
            numbers (list[int]): A list of numbers to store as an attribute in the class.
        """
        self.numbers = numbers

    def add_value(self, value):
        """Adds the value to every number in the numbers attribute, modifying numbers and returning nothing.

        Args:
            value (int): The value to add to each number.
        """
        self.value = value
        
        output = []
    
        for i in numbers:
            output.append(i + value)
        return output

numbers = [1, 2, 3]
value = 5
test = ListOfNumbers(numbers)
test.add_value(value)
test

&#9989; **Question 5.4 What are the advantages and disadvantages of each programming paradigm? (3 points):** 

with linear programming, you can do it quickly and it's more useful for less complicated pieces of code (like simple math). This problem would be a good example of what linear programming is good for. With more complicated problems, however, you would need to rewrite the code numerous times to get the result you want

functional programming can be good for when you're repeating a calculation multiple times for different vaues (so creating a function to solve a problem and then plugging in different values). It gets difficult when it needs to be a more complicated problem, as functions can get messy and hard to read.

object programming is good for complicated problems where there are multiple subsets (classes) of things. you can reuse code a lot and it's easier to maintain. It does, however, take a lot of time to develop and debug.

---
### &#128721; STOP (1 point)
**Pause to commit your changes to your Git repository!**

Take a moment to save your notebook, commit  and push the changes to your Git repository.

---

## You're done! Congrats on finishing your CMSE 202 Midterm!

Make sure all of your changes to your repository are committed and pushed to GitHub. Also upload a copy of this notebook to the dropbox on D2L in case something went wrong with your repository or if you couldn't get the repository to work.